### ISSUE TITLE
Data race in Tier.go Announce method

### ISSUE THREAD

**Issue by bsergean** [cenkalti/rain#25]
I have 2 sessions objects, and I run a download for both.

Here is the race report.

```
==================
WARNING: DATA RACE
Read at 0x00c000666678 by goroutine 81:
  github.com/cenkalti/rain/internal/tracker.(*Tier).Announce()
      /var/lib/jenkins/go/pkg/mod/github.com/cenkalti/rain@v1.7.0/internal/tracker/tier.go:27 +0xb9
  github.com/cenkalti/rain/internal/announcer.announce()
      /var/lib/jenkins/go/pkg/mod/github.com/cenkalti/rain@v1.7.0/internal/announcer/announce.go:24 +0x18f
  github.com/cenkalti/rain/internal/announcer.(*PeriodicalAnnouncer).announce()
      /var/lib/jenkins/go/pkg/mod/github.com/cenkalti/rain@v1.7.0/internal/announcer/periodic.go:244 +0x1fc
  github.com/cenkalti/rain/internal/announcer.(*PeriodicalAnnouncer).doAnnounce·dwrap·5()
      /var/lib/jenkins/go/pkg/mod/github.com/cenkalti/rain@v1.7.0/internal/announcer/periodic.go:238 +0x6e

Previous write at 0x00c000666678 by goroutine 151:
  github.com/cenkalti/rain/internal/tracker.(*Tier).Announce()
      /var/lib/jenkins/go/pkg/mod/github.com/cenkalti/rain@v1.7.0/internal/tracker/tier.go:29 +0x1c4
  github.com/cenkalti/rain/internal/announcer.announce()
      /var/lib/jenkins/go/pkg/mod/github.com/cenkalti/rain@v1.7.0/internal/announcer/announce.go:24 +0x18f
  github.com/cenkalti/rain/internal/announcer.(*PeriodicalAnnouncer).announce()
      /var/lib/jenkins/go/pkg/mod/github.com/cenkalti/rain@v1.7.0/internal/announcer/periodic.go:244 +0x1fc
  github.com/cenkalti/rain/internal/announcer.(*PeriodicalAnnouncer).doAnnounce·dwrap·5()
      /var/lib/jenkins/go/pkg/mod/github.com/cenkalti/rain@v1.7.0/internal/announcer/periodic.go:238 +0x6e

Goroutine 81 (running) created at:
  github.com/cenkalti/rain/internal/announcer.(*PeriodicalAnnouncer).doAnnounce()
      /var/lib/jenkins/go/pkg/mod/github.com/cenkalti/rain@v1.7.0/internal/announcer/periodic.go:238 +0x110
  github.com/cenkalti/rain/internal/announcer.(*PeriodicalAnnouncer).Run()
      /var/lib/jenkins/go/pkg/mod/github.com/cenkalti/rain@v1.7.0/internal/announcer/periodic.go:209 +0x10ec
  github.com/cenkalti/rain/torrent.(*torrent).startNewAnnouncer·dwrap·45()
      /var/lib/jenkins/go/pkg/mod/github.com/cenkalti/rain@v1.7.0/torrent/torrent_start.go:107 +0x39

Goroutine 151 (finished) created at:
  github.com/cenkalti/rain/internal/announcer.(*PeriodicalAnnouncer).doAnnounce()
      /var/lib/jenkins/go/pkg/mod/github.com/cenkalti/rain@v1.7.0/internal/announcer/periodic.go:238 +0x110
  github.com/cenkalti/rain/internal/announcer.(*PeriodicalAnnouncer).Run()
      /var/lib/jenkins/go/pkg/mod/github.com/cenkalti/rain@v1.7.0/internal/announcer/periodic.go:156 +0x284
  github.com/cenkalti/rain/torrent.(*torrent).startNewAnnouncer·dwrap·45()
      /var/lib/jenkins/go/pkg/mod/github.com/cenkalti/rain@v1.7.0/torrent/torrent_start.go:107 +0x39
==================
```

If I read this right, the problem is in this file

```
// Announce a torrent to the tracker.
// If annouce fails, the next announce will be made to the next Tracker in the tier.
func (t *Tier) Announce(ctx context.Context, req AnnounceRequest) (*AnnounceResponse, error) {
    resp, err := t.Trackers[t.index].Announce(ctx, req) // line 27
    if err != nil {
        t.index = (t.index + 1) % len(t.Trackers) // line 29
    }
    return resp, err
}
```

on line 27 and 29 we are reading/writing t.index.
Maybe a fix would be to make t.index an atomic int, and use atomic operations to get/set it ?

Or maybe the problem is with the t.Trackers array ? I will try to something obvious (like adding a mutex to Tier) to debug this further.

**Comment by cenkalti** [comment]
How did you find this out? Can you share your unit test please? It'll be useful to add this test to the project. Thank you.

**Comment by bsergean** [comment]
I’ll try to repro this and make a standalone example.

Now I only use one session, but at some point I did use 2 sessions.

**Comment by cenkalti** [comment]
Fixed in #26

**Comment by bsergean** [comment]
Thanks !
